[visionOS] Fullscreen element doesn't lay out correctly when WKWebView's scale is not initially 1

mwyrzykowski opened the pull request:
#### 6f57d245d7331fb90a87117b006ee6778107408e
<pre>
[visionOS] Fullscreen element doesn&apos;t lay out correctly when WKWebView&apos;s scale is not initially 1
<a href="https://bugs.webkit.org/show_bug.cgi?id=260031">https://bugs.webkit.org/show_bug.cgi?id=260031</a>
&lt;radar://113650540&gt;

Reviewed by Aditya Keerthi.

Set the viewScale to 1 in fullscreen after saving the web view state.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController enterFullScreen:]):

Canonical link: <a href="https://commits.webkit.org/266822@main">https://commits.webkit.org/266822@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad87ef777c75fcdda2a117509b96720d7bfd140e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14787 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15092 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15444 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16534 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13926 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17613 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15192 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14967 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15464 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12559 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17268 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12746 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20321 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13822 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13509 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16742 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14053 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11881 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13344 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3587 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17677 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13896 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -871,6 +871,9 @@ - (void)enterFullScreen:(CGSize)videoDimensions
     page->setSuppressVisibilityUpdates(true);
 
     _viewState.store(webView.get());
+#if PLATFORM(VISION)
```
In the interest of having less platform-specific code, would it be harmful to do this on other platforms as well?

mwyrzykowski replied:
Good idea, it shouldn't be harmful, I will confirm.

I am also interested if anyone knows why it does not seem to be necessary on iPad.

mwyrzykowski replied:
Confirmed it is fine on iOS and actually I think the issue is present there, but we perform another layout and that corrects the layout.

I removed the change from `PLATFORM(VISION)`

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -871,6 +871,9 @@ - (void)enterFullScreen:(CGSize)videoDimensions
 
     _viewState.store(webView.get());
+#if PLATFORM(VISION)
+    [webView _setViewScale:1.f];
```
Do we really want to set the scale before taking the snapshot?

mwyrzykowski replied:
This change isn't needed. It is fine to set the scale below where we set mEDW. I'll remove this change

pxlcoder reviewed (commented):
I'm curious why `WebKit::WKWebViewState().applyTo(webView.get());` doesn't already take care of this.

mwyrzykowski commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -894,6 +894,7 @@ - (void)enterFullScreen:(CGSize)videoDimensions
         [webView setAutoresizingMask:(UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight)];
         [webView setFrame:[_window bounds]];
         [webView _setMinimumEffectiveDeviceWidth:0];
+        [webView _setViewScale:1.f];
```
@pxlcoder I think we only need to set the scale here. I see a similar issue on iOS.

I'm going to confirm if this fixes the issue on iOS too

mwyrzykowski committed:
[visionOS] Fullscreen element doesn't lay out correctly when WKWebView's scale is not initially 1
https://bugs.webkit.org/show_bug.cgi?id=260031
<radar://113650540>

Reviewed by Aditya Keerthi.

Set the viewScale to 1 in fullscreen after saving the web view state.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController enterFullScreen:]):

Canonical link: https://commits.webkit.org/266822@main
